### PR TITLE
merge providers into TickTockZoneRulesProvider

### PR DESCRIPTION
**Pull request by gabrielittner**
`TickTockZoneRulesProvider` is the former `LazyZoneRulesProvider` without the reading of the dat files. That functionality is now extracted and represented by the new `ZoneDataProvider` interface. `ZoneDataLoader` is not needed anymore in the plugins, because it's only used by the data provider which the user instantiates (for the Android artifacts: user = artifact). 

For the classic tzdb files there is a single class that implements both `ZoneDataProvider` and `ZoneIdsProvider` because both are loaded from the same source. The code that reads the stream is a slightly simplified version of the one from threetenbp. I removed support for having multiple versions (e.g. 2019c and 2020a) in one `tzdb.dat` file. In theory we could also simplify the binary format a bit but we would only save 2 shorts per zone, so that's just ~2,4kb

### CODE REVIEW

File: ticktock-android/tzdb/build.gradle [hunk 1]
```diff
@@ -19,13 +19,20 @@ plugins {
   id 'dev.zacsweers.ticktock'
 }
 
+android {
+  compileOptions {
+    coreLibraryDesugaringEnabled true
+  }
+}
```
Comment (ZacSweers) [inline ticktock-android/tzdb/build.gradle:26]: My understanding is that this shouldn't be applied on library projects

Comment (gabrielittner) [reply ticktock-android/tzdb/build.gradle:26]: Lint was failing without because I call `Supplier.get()` in here

Comment (gabrielittner) [reply ticktock-android/tzdb/build.gradle:26]: I can also ignore the lint

Comment (ZacSweers) [reply ticktock-android/tzdb/build.gradle:26]: Yeah I suspect that’s a false positive because of the set minSdk version. Ignoring sounds good 👍

**Pull request merged**